### DISCUSSION TITLE
How to backup SQLite DB with async engine?

### DISCUSSION THREAD

**Discussion by vunhatchuong** [sqlalchemy/sqlalchemy#11936] [Usage Questions] [answered]
I want to backup my SQLite db through `conn.backup()` by using the engine.

This is my session.py

```python
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine
from sqlalchemy.pool import AsyncAdaptedQueuePool, NullPool
from sqlmodel.ext.asyncio.session import AsyncSession

engine = create_async_engine(
    DATABASE_URL,
    echo=False,
    poolclass=AsyncAdaptedQueuePool
)

async_session = async_sessionmaker(
    autocommit=False,
    autoflush=False,
    bind=engine,
    class_=AsyncSession,
    expire_on_commit=False,
)
```

**Comment by zzzeek** [comment]
you want to use the form at https://docs.sqlalchemy.org/en/20/faq/connections.html#accessing-the-underlying-connection-for-an-asyncio-driver and you'd be calling this method: https://github.com/omnilib/aiosqlite/blob/463a2e8e0f7d6eefdc1f449c6f22b211a58c3d03/aiosqlite/core.py#L342

**Reply by vunhatchuong** [reply, marked as answer]
Thanks for the suggestion, what do you think about this way of handling it:

```python
async def backup_sqlite_db(async_engine: AsyncEngine, filename: str) -> None:
    backup_path = Path(DB_BACKUP_DIR)
    if not backup_path.exists():
        await startup()

    backup_path = backup_path.joinpath(filename)

    async with async_engine.connect() as async_conn:
        conn = async_conn.sync_connection
        if conn is not None:
            conn = conn.connection.driver_connection

        with sqlite3.connect(backup_path) as backup_conn:
            if conn is not None:
                await conn.backup(backup_conn)
```

**Reply by zzzeek** [reply]
that's fine.  you can likely skip the "if conn is not None" part since it wont be None and you wouldn't want the routine silently doing nothing in any case if it were.

**Reply by CaselIT** [reply]
instead of `sqlite3` you will likely want to use `aiosqlite` unless backup requires a sync sqlite3 connection

**Reply by vunhatchuong** [reply]
Thank you guys for the help.
